model-mesh-ovms-adapter/server/modelmanager.go

```diff
@@ -261,86 +260,19 @@ func (mm *OvmsModelManager) run() {
-		// period of time after a request is received
-		batch := make([]*request, 0, mm.config.BatchPreallocSize)
-		var stopChan <-chan time.Time
-
-	runLoopSelect:
-		select {
-		case req, ok := <-mm.requests:
-			if !ok {
-				// shutdown the loop after processing this batch
-				mm.requests = nil
-				break // from select
-			}
-			// add request to batch
-			batch = append(batch, req)
-
-			// set the stop timer, if not set already
-			if stopChan == nil {
-				stopChan = time.NewTimer(mm.config.BatchWaitTime).C
-			}
-
-			goto runLoopSelect
-		case <-stopChan:
-			break // from select
-		}
-
-		log.V(1).Info("Processing batch of requests", "numRequests", len(batch))
-
-		// update the model configuration based on the requests
-		var unloadAllRequest *request
-
-		// build a new map to track changes to the same model
-		modelUpdates := map[string]*request{}
-		for _, req := range batch {
-			// check for an unloadAll
-			if req.requestType == unloadAll {
-				if unloadAllRequest != nil {
-					// unexpected situation
-					log.Info("Got multiple UnloadAll requests in one batch")
-					completeRequest(unloadAllRequest, codes.Aborted, "Subsequent UnloadAll request received")
-				}
-
-				unloadAllRequest = req
-				// abort any requests preceding the unloadAll
-				for id, req := range modelUpdates {
-					completeRequest(req, codes.Aborted, "Model Server is being reset")
-					// remove the pointer from the updates map
-					delete(modelUpdates, id)
-				}
-				continue
-			}
-
-			if modelUpdates[req.modelId] != nil {
-				// abort the prior request first
-				completeRequest(modelUpdates[req.modelId], codes.Aborted, "Concurrent request received for the same model")
-			}
-
-			modelUpdates[req.modelId] = req
-		}
+		loadRequestsMap := mm.gatherLoadRequests()
 
-		// if the batch included an UnloadAll, reset the model config map prior adding the updates
-		if unloadAllRequest != nil {
-			log.V(1).Info("Processing UnloadAll", "numRequests", len(batch))
-			mm.loadedModelsMap = map[string]OvmsMultiModelConfigListEntry{}
-		}
+		// gatherLoadRequests() collects requests over time, some requests
+		// may have been cancelled by now. Check that here before
+		// reloading the config
+		for id, req := range loadRequestsMap {
+			if err := req.ctx.Err(); err != nil {
+				s := status.FromContextError(err)
+				mm.log.V(1).Info("Aborting request with cancelled context before reloading", "model_id", id, "request_type", req.requestType, "code", s.Code())
+				completeRequest(req, s.Code(), "Request context has been cancelled")
 
-		// process the updates
-		for id, req := range modelUpdates {
-			switch req.requestType {
-			case load:
-				mm.loadedModelsMap[id] = OvmsMultiModelConfigListEntry{
-					Config: OvmsMultiModelModelConfig{
-						Name:     req.modelId,
-						BasePath: req.basePath,
-					},
-				}
-			case unload:
 				delete(mm.loadedModelsMap, id)
+				delete(loadRequestsMap, id)
```

Comment:
What if the model was already loaded (in the loaded map) prior to this load request?

I know that's a weird / probably-would-never-happen case and that model-mesh would send an unload anyhow. But I'm thinking again about semantics of "if request fails state should ideally remain unchanged".

One way to tweak the current logic would be to add a previousState field to the request struct which stores a `OvmsMultiModelConfigListEntry` ptr (usually `nil`). Then if replacing/aborting would just copy from the old request into the new one.

Reply:
Hmm, well we get here after the storage has already been processed, and there is no reversion of the file changes. Would it still be worth keeping the old model loaded?
I would expect that ModelMesh does not send inference request to a model that has failed a load request so even if it doesn't send an Unload, it would need to complete a Load before the model is used again.

If you do want reversion to previous state on load failure, should we revert state if a model load fails after the reload as well? What about in the case that the request is cancelled during the reload (regardless of whether or not the load succeeded)?

Reply:
Yeah good point about the storage. I guess if we were to do this properly we would process the requests before the storage step and loads of already-loaded models if the parameters are different / complete them immediately if params match.

So maybe best/fine to leave as-is then.